src/main/java/com/conveyal/r5/profile/FastRaptorWorker.java

```diff
@@ -323,74 +348,61 @@ private void advanceScheduledSearchToPreviousMinute (int nextMinuteDepartureTime
+                // Perhaps we should only do it when iterationsPerMinute is high (2 or more?)
+                if (transit.hasFrequencies && boardingMode == MONTE_CARLO && ENABLE_OPTIMIZATIONS) {
                     raptorTimer.scheduledSearchFrequencyBounds.start();
                     doFrequencySearchForRound(scheduleState[round - 1], scheduleState[round], UPPER_BOUND);
                     raptorTimer.scheduledSearchFrequencyBounds.stop();
                 }
-
+                // Apply transfers to the scheduled result that will be reused for the previous departure minute.
+                // Transfers will be applied separately to the derived frequency result below, when relevant.
                 raptorTimer.scheduledSearchTransfers.start();
                 doTransfers(scheduleState[round]);
                 raptorTimer.scheduledSearchTransfers.stop();
             }
             raptorTimer.scheduledSearch.stop();
         }
 
-        // Now run frequency searches using randomized schedules for all frequency lines. We use the scheduled search
-        // and the worst-case boarding wait for all frequency routes to establish an upper bound on arrival times for
-        // the frequency searches, so we are copying the arrival times from the just completed search.
+        // Now run one or more frequency searches using randomized schedules for all frequency lines.
+        // These will improve upon any upper bound travel times just established at this departure minute.
         // This is a key innovation, described in more detail in Conway, Byrd, and van der Linden 2017.
+        // Unlike those scheduled results, these frequency results are not reusable at the previous minute because
+        // the schedule is almost certain to be different at that previous minute.
         if (transit.hasFrequencies) {
             raptorTimer.frequencySearch.start();
             int[][] result = new int[iterationsPerMinute][];
+
             // In Monte Carlo mode, each iteration is a fresh randomization of frequency route offsets.
             // In half-headway mode, only one iteration will happen and schedules will not be randomized.
             for (int iteration = 0; iteration < iterationsPerMinute; iteration++) {
-                // For each new randomized schedule, we start with a fresh copy of the upper bound travel times.
-                RaptorState[] frequencyState = Stream.of(scheduleState)
-                        .map(RaptorState::copy)
-                        .toArray(RaptorState[]::new);
-
-                // Recreate the chain of pointers between rounds for the freshly copied states.
-                for (int i = 1; i < frequencyState.length; i++) {
-                    frequencyState[i].previous = frequencyState[i - 1];
-                }
-
+                // Make a fresh copy of the upper bound travel times for each new randomized schedule (iteration).
```

Comment:
A comment in `McRaptorSuboptimalPathProfileRouter` ("we're not using range-raptor so it's safe to change the schedule on each search") could be read to imply that it's not safe to randomize the offsets on each search. That could be clarified with respect to my comment a few lines up.

Reply:
I don't fully understand this comment, let's discuss at today's meeting.

Reply:
Confirmed in discussion that indeed McRaptor does not apply the range raptor optimization.